[ENH] Apply forecasters depending on series type (e.g., intermittent) compositor

shlok191 opened the pull request:
#### Reference Issues/PRs
<!--Implements a compositor which accepts a series-to-primitives transformer as a parameter; fixes #6279 -->

#### What does this implement/fix? Explain your changes.
This implementation introduces the `TransformSelectForecaster` which accepts a series-to-primitives transformer and a dictionary of forecasters as initialization parameters. This feature will allow users to categorize their input data through a transformer and then apply specific forecasters best suited for the classification categories.

#### What should a reviewer concentrate their feedback on?

A reviewer should most focus on the `_category_compositor.py` file in the `sktime/forecasting/compose` directory! It might also be helpful to take a look at the `ADICVTransformer` as it is a default parameter for this compositor.

#### Did you add any tests for the change?

I have not implemented any tests yet, but I am planning on adding them once I can get more feedback about the implementation and some advice on good test cases for this particular function!

#### PR checklist


##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [x] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [x] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [x] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [x] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

shlok191 commented:
@fkiraly, Hello Franz, I've made the changes that we discussed prior! I have made the series-to-primitives transformer a passable parameter and attempted to generalize the model. I had two questions I wanted to discuss though:

- For the passed transformer, is it safe for me to assume that the first column from a `transform` function will always include the data I require to choose a forecaster?

- For the test cases, would you have an idea about what might be good scenarios? I can think about instances that might lead to clear failures (like not passing in all required forecasters) but I think there can be better ones! 

Thanks a lot for the help!

fkiraly commented on sktime/forecasting/compose/_category_compositor.py:
```diff
@@ -0,0 +1,260 @@
+        self : reference to self
+        """
+        # reset the state of the forecaster
+        self.reset()
```
input checks and logic does not need to be carried out here. This is done in the boilerplate layer, in `BaseForecatser.fit`.

shlok191 replied:
Ah, got it. Removing the checks from all relevant places!

fkiraly replied:
have you done that? still seems like they are there

shlok191 replied:
I was at work at the time, but I've made the changes now

fkiraly commented on sktime/forecasting/compose/_category_compositor.py:
```diff
@@ -0,0 +1,260 @@
+            Point predictions
+        """
+        # Check if we have a fited forecaster to the data
+        self.check_is_fitted()
```
same

fkiraly commented on sktime/forecasting/compose/_category_compositor.py:
```diff
@@ -0,0 +1,260 @@
+            self.category_ in self.forecasters
+        ), f"Forecaster not provided for given time series of type {self.category_}."
+
+        self.chosen_forecaster_ = self.forecasters[self.category_]
```
makes sense, I wonder whether we can make it more robust by allowing the user to specify a `fallback_forecaster`, e.g., if the category is sth else. I would make it the third, optional argument, and use it as a default in a dict.get if it is not `None` - otherwise, raise a `ValueError` if no fallback is provided and the category is not found.

shlok191 replied:
I think that's definitely a good idea, I'll add it to the code :)

shlok191 commented on sktime/forecasting/compose/_category_compositor.py:
```diff
@@ -0,0 +1,260 @@
+        return self.predict(fh, X)
+
+    @classmethod
+    def get_test_params(cls, parameter_set="default"):
```
Adding in test cases!

fkiraly commented on sktime/forecasting/compose/_category_compositor.py:
```diff
@@ -0,0 +1,260 @@
+        assert y is not None
+
+        # passing time series through the provided transformer!
+        self.category_ = self.transformer.fit_transform(y=y).iloc[0, 0]
```
Should this not be `X=y`?

shlok191 replied:
Ah, made this fix in the new commit!

fkiraly reviewed (changes requested):
You need to make sure the estimator is exported in a public module, or tests will not be run. E.g., import to `compose.__init__.py`.

This should also highlight other issues - you can do the same by running `check_estimator`.

shlok191 commented:
@fkiraly, I am running the pytests locally and did not encounter the issues I'm seeing online. Could you please tell me if I'm doing something wrong with the soft dependencies? Thanks!

fkiraly commented on sktime/forecasting/compose/tests/test_category_compositor.py:
```diff
@@ -0,0 +1,108 @@
+        (
+            {
+                "smooth": NaiveForecaster(),
+                "intermittent": SARIMAX(),
```
this requires a soft dependency, how about you use `Croston` instead, which is better for intermittent demand anyway?

fkiraly commented on sktime/forecasting/compose/tests/test_category_compositor.py:
```diff
@@ -0,0 +1,108 @@
+        (
+            {
+                "smooth": NaiveForecaster(),
+                "intermittent": SARIMAX(),
```
same

fkiraly commented on sktime/forecasting/compose/_category_compositor.py:
```diff
@@ -0,0 +1,252 @@
+            "forecasters": {
+                "smooth": NaiveForecaster(),
+                "erratic": PolynomialTrendForecaster(),
+                "intermittent": SARIMAX(),
```
this requires a soft dependency, how about you use `Croston` instead, which is better for intermittent demand anyway?

fkiraly commented on sktime/forecasting/compose/_category_compositor.py:
```diff
@@ -0,0 +1,252 @@
+class CategoryCompositor(BaseForecaster):
+    """Compositor that utilizes varying forecasters by time series data's nature.
+
+    Applies a series-to-primitives transformer on a given time series and utilizes
```
can you be more precise regarding what this does?

fkiraly commented on sktime/forecasting/compose/_category_compositor.py:
```diff
@@ -0,0 +1,252 @@
+    Raises
+    ------
+    AssertionError: If a valid transformer (an instance of BaseTransformer)
+    is not passed or if valid forecasters (instances of BaseForecaster) is not given.
```
an example would be nice - you could simply code from one of the test cases

fkiraly commented on sktime/forecasting/compose/_category_compositor.py:
```diff
@@ -0,0 +1,252 @@
+        "y_inner_mtype": "pd.Series",
+        "X_inner_mtype": "pd.Series",
+        "scitype:y": "univariate",
+        "ignores-exogeneous-X": True,
```
tags should be set differently - I would set scitype:y to both, ignores X to False; the inner mtypes to `pd.DataFrame` to handle multivariate data.

the capabilities, including "handles missing data" should be set so a capability is true if and only if all forecasters have the capability.

There is another thing we should do once this is sorted and tests pass, to handle global forecasters and hierarchical series - let´s make a note for it now, and I will explain it later once the simple case works.

fkiraly commented:
> I am running the pytests locally and did not encounter the issues I'm seeing online. Could you please tell me if I'm doing something wrong with the soft dependencies? Thanks!

Which failures precisely, can you describe?

fkiraly commented:
Re failures currently, I would suggest you use `check_estimator` to debug them, see guide here:
https://www.sktime.net/en/latest/developer_guide/add_estimators.html#using-the-check-estimator-utility

Feel free to ask in the `dev-chat` discord channel if you get stuck

shlok191 commented:
@benHeid, @astrogilda Hey guys! I'm nearly done working on this PR.

I've received some great feedback from Franz which I've tried to incorporate, and I'll really appreciate it if y'all could take a look at my code as well. 

Thank you so much!

fkiraly commented:
Sure! short remark, the "protocol" for pull request review is that you should comment on the open discussions, one of:

* explain how addressed
* argue that needs not be addressed
* request information or discussion

Until consensus is reached.

fkiraly commented on sktime/forecasting/compose/_category_compositor.py:
```diff
@@ -0,0 +1,479 @@
+adi_cv_transformer = ADICVTransformer(features=["class"])
+
+
+class CategoryCompositor(BaseForecaster):
```
I think the name is a bit non-self-explanatory.

How about `ForecastByGroup`, or `ByGroupForecaster`?

shlok191 replied:
I like the sound of `ForecastByGroup`, I'll change the name!

fkiraly commented on sktime/forecasting/compose/_category_compositor.py:
```diff
@@ -0,0 +1,479 @@
+    def __init__(
+        self,
+        forecasters,
+        transformer=adi_cv_transformer,
```
mutable objects should never be used as defaults in python methods or classes!

The reason is, if the object is mutated, then the default for any future calls is also mutated.

The pythonic way to do is is to have the defalut = `None`, and then check if is `None`, then set the param to the default.

Now, in `sklearn` API, params should be written to self and never overwritten, so you ought to write the default or transformer to `self.transformer`.

shlok191 replied:
Ah, got it. That makes sense. I'll ammend this and add it to the next commit.

fkiraly commented on sktime/forecasting/compose/_category_compositor.py:
```diff
@@ -0,0 +1,479 @@
+        self.forecasters = forecasters
+        self.fallback_forecaster = fallback_forecaster
+
+        self.transformer_ = copy.deepcopy(transformer)
```
don't copy, but do `transformer.clone()`.

shlok191 replied:
Got it, I'll have this change in the next commit.

fkiraly commented on sktime/forecasting/compose/_category_compositor.py:
```diff
@@ -0,0 +1,479 @@
+            "X-y-must-have-same-index": False,
+            "enforce_index_type": False,
+            "handles-missing-data": False,
+            "capability:insample": False,
```
these should be set to `True`, to signifiy that it might be, depending on the components used.

shlok191 replied:
Got it, I've set these to true for both `true_if_any_tags` and `true_if_all_tags`.

fkiraly commented on sktime/forecasting/compose/_category_compositor.py:
```diff
@@ -0,0 +1,479 @@
+
+        # Extrapolating values for flags that should be True if they are
+        # True for all of the given forecasters!
+        for tag in true_if_all_tags.keys():
```
this block of code is very long. Why not use `self._anytagis_then_set`?

shlok191 replied:
I actually tried out `self._anytagis_then_set`, but I got the error that the function was not defined. Is it a defined function that the class should have inherited from `BaseForecaster`?

fkiraly commented on sktime/forecasting/compose/_category_compositor.py:
```diff
@@ -0,0 +1,479 @@
+
+            # Adopt the fallback forecaster if possible
+            else:
+                self.chosen_forecaster_ = copy.deepcopy(self.fallback_forecaster)
```
`clone`, not deepcopy

fkiraly commented on sktime/forecasting/compose/_category_compositor.py:
```diff
@@ -0,0 +1,479 @@
+        """
+        # passing time series through the provided transformer!
+
+        self.category_ = self.transformer_.fit_transform(X=y).iloc[0, 0]
```
this should be explained in the docstring, i.e., that the first column is used.

Or, do we want to support multiple columns? In which case the keys could perhaps be tuples? You can get these via `itertuples`, and the length 1 case needs special treatment.

shlok191 replied:
We could do it that way. However, would it then be a requirement that the first row must have the key that we are looking for? Also, if we are doing multiple columns (by which I assume you mean that the key could be in any column, but it would be in only in one specific column) would it be simpler if the user could pass an additional column name as a parameter where we should search for the key?

I'm okay with either design choice (user possibly specifies which column to choose or sets up the transformer such that the first column is the valid one), but could you tell me which approach might be easiest for the user?

fkiraly replied:
I would say, let's use the first column for now and worry about more columns later.

fkiraly commented on sktime/forecasting/compose/_category_compositor.py:
```diff
@@ -0,0 +1,479 @@
+        -------
+        self : reference to self
+        """
+        self.chosen_forecaster_.update(y=y, X=X, update_params=update_params)
```
this should also be explained in the docstring, i.e., that the transformation is not updated.

or, we could make it a parameter whether the transformation updates, but perhaps not in this PR (seems fiddly)

shlok191 replied:
I agree that it might be good to add that as a parameter (if the transformer updates, and be extension the chosen forecaster updates too). I'll explain this in the docstring for now but let me know if you'd like me to add this in for this PR!

fkiraly commented on sktime/forecasting/compose/_category_compositor.py:
```diff
@@ -0,0 +1,479 @@
+    >>> from sktime.transformations.series.adi_cv import ADICVTransformer
+    >>> from sktime.transformations.series.tests.test_adi_cv import (
+    ...     _generate_erratic_series)
+    >>> category_compositor = CategoryCompositor(
```
nice example! would you like to add comments or explanation what this does? it is not obvious for the user, and a good example is accompanied by a good explanation (comments or text)

shlok191 replied:
Ah, absolutely! I'll explain it in more detail. Sorry if this seems rushed, I've just been reading the LagLlama paper and have been very excited about getting started on adding that implementation as soon as I could... 😅

fkiraly reviewed (changes requested):
Thanks, really great contribution!

I've reviewed the estimator, and the main thing I want to bring up: the forecasters for the components are not tunable via `get_params` etc. For that, we would have to use `_HeterogenousMetaEstimator`, which adds that feature - and also the "any tag" method that is missing.

The best template to use here would be `HierarchyEnsembleForecaster`, imo.

To use the ensemble base class, we may also have to change the `dict` syntax to list of `("name", forecaster, value)`, so the ensemble can be used.

Let me know what you think and whether this makes sense.

shlok191 commented:
@fkiraly, instead of using `HierarchyEnsembleForecaster` for which I am a little unsure the changes I might have to make to ensure that `_fit` works, could I possibly try to extend `_HeterogenousMetaEstimator`? I have read `HierarchyEnsembleForecaster`'s internal functions to assist with `get_params` and I think it might be easier for me to implement them for this forecaster!

fkiraly commented:
> could I possibly try to extend `_HeterogenousMetaEstimator`

That's what I meant! Sorry if my phrasing was confusing as to the precise architecture.

I meant, the `HierarchyEnsembleForecaster` is a good example on how `_HeterogenousMetaEstimator` is meant to be used as an intermediate base class.

shlok191 commented:
Ah, got it! I thought I’d confirm. As a short update, I’ve completed the implementation locally and just have some bugs to fix to pass all the test cases. This should be done soon!

fkiraly commented:
ready for review?

shlok191 commented:
@fkiraly, yes! I have tried to follow the lead from `HierarchyEnsembleForecaster` to implement the necessary functionality to ensure get_params and other related functions work as expected.

fkiraly commented on docs/source/api_reference/forecasting.rst:
```diff
@@ -50,6 +50,7 @@ Pipelines can also be constructed using ``*``, ``+``, and ``|`` dunders.
     ForecastByLevel
     Permute
     HierarchyEnsembleForecaster
+    CategoryCompositor
```
old name still here

shlok191 replied:
Sorry, I missed that, removing

fkiraly commented on sktime/forecasting/compose/_by_group_forecaster.py:
```diff
@@ -0,0 +1,526 @@
+from sktime.transformations.series.adi_cv import ADICVTransformer
+
+
+class ByGroupForecaster(BaseForecaster, _HeterogenousMetaEstimator):
```
the name still feels a bit uninformtaive to me.

It should imply that the groups are computed from the time series. Not sure how to do that, that might be a good question to brainstorm.

shlok191 replied:
I agree. I wanted to have a name that captures the 2-stage nature of this with something that could showcase that there is a `series-to-primitives` transformer and then a forecaster but I just cannot think of a name that is not a mouthful.

Should I drop this question in dev-chat? Maybe someone has some good ideas

shlok191 replied:
How about something like `TransformSelectForecaster` or `PrimitivesDerivedForecaster`?

shlok191 replied:
Sounds good! I'm planning on pushing in a new commit with the added changes soon and with the name `TransformSelectorForecaster` and I'll request folks to pitch in ideas on Discord. If we get something new I'll add it in!

Also, yes. I'll add an example too when I write the message!

fkiraly commented on sktime/forecasting/compose/_by_group_forecaster.py:
```diff
@@ -0,0 +1,526 @@
+
+# Function implementations that will be added dynamically
+# if the conditions are met. explained further above!
+def _update_predict_single(self, y, fh, X=None, update_params=True):
```
is this redundant? Why is this needed?

shlok191 replied:
Right, there isn't exactly much of an efficient improvement there. I'll remove this

fkiraly commented on sktime/forecasting/compose/_by_group_forecaster.py:
```diff
@@ -0,0 +1,526 @@
+            The list of forecasters which is returned. Also includes the
+            fallback forecaster with the category: "fallback_forecaster"
+        """
+        return list(self.forecasters.values()) + [
```
should this be `items` rather than `values`?

shlok191 replied:
Oh, that's right! I'll make the change

fkiraly replied:
I find it odd that the tests did not detect this, how about adding an extra test that checks that the `get_params` keys are as expected?

Further, I think you may want to check whether you have to coerce the keys to `str`, or whether it works without giving trouble (because this happens somewhere else already).

shlok191 replied:
Yes, I'm adding in that check with the next commit. Could you explain what you mean by coercing the keys to string? Is that in reference to the `"fallback_forecaster", self.fallback_forecaster` line where I specify fallback forecaster as a string key?

fkiraly replied:
If you run `get_params` with `deep=True`, nested parameter keys will be strings `component__parametername`.

For the `_HeterogeneousMetaEstimator` to work, the name in name/estimator tuples needs to be string.

In cases where the transformer outputs non-string values (e.g., integers), in the property you may have to coerce the integers to strings - or maybe it is handled already somewhere.

shlok191 commented:
@fkiraly, I have added in the changes we discussed! I have added tests to ensure that the parameters return from `get_params()` are as we hope they should be and that the `name` in the `(name, object)` tuples is are always strings.

Additionally, I have renamed the forecaster to `TransformSelectForecaster` and I am dropping a question in dev-chat about the update!

fkiraly commented on sktime/forecasting/compose/_transform_select_forecaster.py:
```diff
@@ -0,0 +1,515 @@
+        """
+        # passing time series through the provided transformer!
+
+        self.category_ = self.transformer_.fit_transform(X=y).iloc[0, 0]
```
I think the args should be `(X=y, y=X)`, in case the transformer takes exogeneous data?

fkiraly replied:
added that, hope it does not break anything

shlok191 replied:
Thank you so much! I just got off work and was just about to add it in. I'll keep an eye on the tests :)
